doc/dev/osd_internals/manifest.rst

```diff
@@ -110,6 +187,49 @@ cache/tiering promotion to ensure that the ObjectStore ends up with
 the right byte range sharing via clone.  We probably want to ensure
 that adjacent clones agree on chunk mappings for shared regions.
 
+* Implementation
+
+  My thought is that set_chunk and set_reredirect shouldn't set
+  user_modify to indicate not to trigger a new clone (probably also need a flag like cache_evict).
+  Also, to use manifest with snapshot, set_chunk should be set when the manifest object is created,
+  and disallow to make a new clone on the manifest object which already has clones.
+  As a result, the overall procedure is:
+  1. Write the object A
+  2. SET_CHUNK (offset: 0 ~ 4)
+  3. SET_CHUNK (offset: 8 ~ 12)
+  4. SET_CHUNK (offset: 16 ~ 20)
+  5. Create a snapshot 
+  6. Write the object A
+  7. Create a snapshot
+  When a snapshot is created between 5 and 6, clones prior to head are dirty
+  if the head object is dirty (and the flush should start from old clones).
+
+  The state of the chunk_map in clone should be MISSING after eviction is done.
+  If we want to read clones, we have to look at the chunk_info_t according to given snap_id 
+  to find out which chunk is needed.
+  Therefore, clone needs to be mutable. To do that, we probably need a new op type.
+  If snap read occurs, the manifest object might have the overwritten chunks because
+  the contents are different depending on what snap_id is.
+  Therefore, flush must happen to avoid the stale data before overwriting the chunks.
```

Comment:
I'm not sure I'm following this.  SET_CHUNK *must* be setting a manifest region that maps to what's already in the object, right?  Clones can't be actually mutated, so other than the case where the head object was dirty, clone extents can't be dirty.

Reply:
I missed something. Can you check the following case is correct? 
Suppose that there is an object that has snapshotted manifest.

Head --> (offset: 0, len: 4, fingerprint oid: aaa, dirty), (offset: 10, len: 4, fingerprint oid: ccc) 
Snapshot A --> (offset: 0, len: 4, fingerprint oid: aaa, dirty), (offset: 10, len: 4, fingerprint oid: ddd) 
Snapshot B --> (offset: 0, len: 4, fingerprint oid: aaa, dirty), (offset: 10, len: 4, fingerprint oid: bbb) 

The second chunk in the head (offset: 10) is mutated, so it has different fingerprint oid: ccc.
And, clone_overlap probably is (offset:0), (offset:10). 
In this case, if we want to read snapshot B, we need to rollback to snapshot B, and then read the head (right?). At this time, we need the oid (bbb) and the oid (bbb) should be stored somewhere. But, where the bbb is recorded? It probably exists in the lower tier (pool), and its oid is "bbb". Therefore, we need to promote it to on the base tier,  then the "bbb" will be overwritten to the head as the content of offset 10 (right?).

Reply:
Regarding clone_overlap, contents of the first chunk in Head, Snapshot A, and Snapshot B are the same and has overlapped position. So, we don't need to drop the reference. But, the second chunks has the different contents even though those are overlapped. Therefore, it's reference will be dropped if dereference occurs.

Reply:
No.  You can read *directly* from snapshots without doing a rollback.  Such reads must return the contents of the object at the time the snapshot was taken.  rbd supports mounting rbd snapshots read-only  cephfs has a similar idea of a read-only mount of a subdirectory at a previous snapshot.  Both of those features rely on this behaviour.

clone_overlap is an optimization that ensures that recovery preserves the underlying ObjectStore level byte range sharing inherent in clone.  Consider:

HEAD:                      [1~10 nochunk] [10~10 nochunk] [20~10 nochunk]
clone 10(9, 7):          [1~10 nochunk] [10~10 nochunk] [20~10 nochunk]
clone 6(6, 5):            [1~10 nochunk] [10~10 nochunk] [20~10 nochunk]
clone 4(2, 1):            [1~10 nochunk] [10~10 nochunk] [20~10 nochunk]

clone_overlap: {10: [1~10, 20~10], 6: [10~10], 4: [1~10, 10~20]}

At this point, no head or clone has any manifest chunks set.  clone_overlap indicates that 10 shares two regions with HEAD, 6 shares 1 with 10, 4 shares 2 with 6.  Suppose we're calling 

SET_CHUNK clone 6 20~10 aaa

In this case, you need to ensure that clone_overlap[4] and clone_overlap[6] no longer consider 20~10 shared with 6.  It might even make sense to always set clone 4's 20~10 region to aaa at the same time since you already know it has the same contents. 

I'd like you to explain in this doc how each of your operations interacts with clone_overlap, both at clone time (when a write operation comes in triggering a clone) and when performed on a clone which already exists.

Reply:
It seems that I need to clear a snap read situation. Current tier_promote() reads the chunks to put together the head object (which has chunked manifest). So, it only contain the chunks for the head. At this point, If we want to read a clone which belongs to the head.  How can we read the clone ? We might have chunk_info_t for the clone, but do not have actual content.


Regarding clone_overlap,
Do we need to consider clone_overlap at clone time?
How about taking clone_overlap into account only when set_chunk is called ?
In your case, 
if SET_CHUNK is called, a calculation needs to occur  in order to get the fingerprint oid, and then adjusts clone_overlap between prior clone and current clone via set_chunk with a flag which trigger immediate calculation.

Reply:
If you take a look at the osd IO path, you'll find that when a snap is specified, find_object_context returns the ObjectContext for the clone rather than for the head.  You'd then look at the manifest on the clone's object context and promote or proxy as appropriate just as you would for head.

For clone_overlap, I think you need to go and have a look at the existing users.  Basically, you need to make sure that if you remove an extent from a clone during SET_CHUNK, you appropriately remove the corresponding portions of clone_overlap so that recovery won't assume that it's already present.

For the initial clone, I *think* it should be fine as long as SET_CHUNK is correct since SET_CHUNK shouldn't generate a clone and normal writes should behave correctly as-is.

Reply:
>If you take a look at the osd IO path, you'll find that when a snap is specified, >find_object_context returns the ObjectContext for the clone rather than for the head. You'd >then look at the manifest on the clone's object context and promote or proxy as >appropriate just as you would for head.

Understand.

>For clone_overlap, I think you need to go and have a look at the existing users. Basically, >you need to make sure that if you remove an extent from a clone during SET_CHUNK, you >appropriately remove the corresponding portions of clone_overlap so that recovery won't >assume that it's already present.

So, in your case

HEAD: [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 nochunk]
clone 10(9, 7): [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 nochunk]
clone 6(6, 5): [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 nochunk]
clone 4(2, 1): [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 nochunk]
clone_overlap: {10: [1 ~ 10, 20 ~ 10], 6: [10 ~ 10], 4: [1 ~ 10, 10 ~ 20]}
SET_CHUNK clone 6 20 ~ 10 aaa

need to become as below
HEAD: [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 nochunk]
clone 10(9, 7): [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 nochunk]
clone 6(6, 5): [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 aaa]
clone 4(2, 1): [1 ~ 10 nochunk] [10 ~ 10 nochunk] [20 ~ 10 aaa]
clone_overlap: {10: [1 ~ 10, 20 ~ 10], 6: [10 ~ 10], 4:[1 ~ 10]}

What I need to do is that make [20 ~ 10 aaa] both clone 4 and 6 due to these has overlap.
And then, remove 4:[20 ~ 10] to avoid that recovery does not assume that it's present ?
In other words set_chunk to clone 6 and clone 4 -> flush clone 6 to store the content -> 
-> remove overlap area between 4 and 6  ?

Reply:
Note, that if there were a clone smaller than 4, you'd have to check that as well, but yes, I think that's right.  I expect that as you test your implementation you'll find other edge cases, however.

Reply:
OK.